cpp/scenario/gazebo/src/World.cpp

```diff
@@ -278,12 +278,16 @@ scenario::core::ModelPtr World::getModel(const std::string& modelName) const
     return pImpl->models[modelName];
 }
 
-bool World::insertModel(const std::string& modelFile,
+bool World::insertModel(const std::string& modelString,
```

Comment:
We've got lucky here that SWIG does not generate `kwargs` for C++ methods with default arguments like this one, otherwise this change would break Python code. Nonetheless, I'm am debating whether a more explicit `World::insertModelFromString()` method would be preferable.

Reply:
Good catch! In a pure-python API this would indeed be non-backward compatible. Then again, we could easily fix it by re-introducing `modelFile` as a kwarg and plumbing it with a deprecation warning.

It might be better, yes. I'm not sure about the mismatch between `insertModelFromString` and `insertModel`; perhaps `insertModelFromString` and `insertModelFromFile`? 

In pure C++ I would have liked to see an overload of the form 

```
bool World::insertModel(const std::string& sdf_string, ...
bool World::insertModel(const std::filesystem::path& sdf_file, ...
```

Unfortunately, `sdf_file` is also a string making it effectively the same function.

In Python, I would have opted for a signature of the form `world.insert_model(self, model_sdf)` where `model_sdf` is an sdf-like object, aka, everything that can be meaningfully converted into sdf. This way, the same signature can cover a large array of inputs, from raw strings, over file paths, to more complex things like streams or fuel URLs (which we can wrap into a minimal sdf and load using the fuel loader, absorbing `getModelFileFromFuel`). This seems cleaner than having a new function for each input type. 

In either case, I'm happy to change, just let me know :)

Reply:
> perhaps `insertModelFromString` and `insertModelFromFile`?

I like the propolal of these explicit methods. In order not to break the APIs, maybe what could be done is providing both new specific methods `*{FromString,FromFile}` and in future `*FromFuel` (I like the idea to move it here), and an overall method `insertModel(std::string, core::Pose, std::string)`  that tries to figure out automatically what the user wants. In edge cases, automatic detection could fail, in this way users can fall back to specific methods.

We can split multiple PRs:

1. This one: move [this code](https://github.com/robotology/gym-ignition/blob/def00f7e47c8ab8e117b2bf77056226501bb7e68/cpp/scenario/gazebo/src/World.cpp#L292-L388) to a new:
   ```cpp
   World::Impl::insertModel(
       const std::string& modelFile, 
       const core::Pose& pose, 
       const std::string& overrideModelName,
       const World& world)  // This is needed just for this->modelNames()
    ```
   and call it from both the existing `World::insertModel` (just from file for the moment) and the new `World::insertModelFromString`
2. Next PR could add a new `World::insertModelFromFile` / `World::insertModelFromFuel` and implement the automatic detection. I remember in the past having problems in multi-os and multi-compiler with `std::filesystem`, few compilers were not happy with their partial implementation, but it was early times. Maybe with the compilers we officially support after the switch to Focal, the situation has improved. Worse case scenario, also `ignition/common/Filesystem.hh` could be used. 

Yesterday I had a look at Edifice and the sdformat11 deprecations, now `sdf::Root` is limited to contain just one `sdf::Model`, therefore the logic of these methods will be simplified with the next release.

Last comment, last week I started thinking about [frame semantics](http://sdformat.org/tutorials?tut=pose_frame_semantics_proposal), and particularly inserting models using `<pose relative_to='frame_name'>0 0 0 0 0 0</pose>`. I originally thought that it could have affected the signature of these `insertModel*` methods, but thinking again now, by extending `core::Pose` maybe we can prevent that and live happy without adding even more methods or breaking existing symbols.

Reply:
> For this PR I would suggest to add both World::insertModelFromFile and World::insertModelFromString. These can exist in parallel with the old World::insertModel without breaking backwards compatibility.

In this case, `World::insertModel` can become a relay of `World::insertModelFromFile`. If you want to integrate it in this PR, ok from my side. Otherwise I can do it in a new one.

> In the second PR we can choose between two options: (1) Depreciate insertModel and only have explicit functions and rely on the user explicitly specifying the conversion. (2) Make insertModel auto-detect the appropriate conversion to sdf (perhaps with a method="auto" overwrite option) and depreciate the explicit methods. It's a philosophical decision on whether or not it makes sense to force the user to be explicit. Depreciation could happen with Fortress and would be in line with the current "under rapid development" label of the library. We could of course keep both, but the redundancy might be confusing to a user that tries to figure out which method to use.

I'd go towards (2), but it's not urgent, Fortress in any case is far away.

> In a third PR we can address World::insertModelFromFuel(std::str url, ...) (or the ability of understanding fuel URLs as input to insertModel) in favor of the current loadModelFromFuel+insertModel. Note that fuel support is already implied if it occurs as an include inside of models loaded from *fromString or *fromFile as soon as #309 is merged.

Yep, what I meant is providing a way into `GazeboSimulator` to do something similar to what today is done by  `utils::getModelFileFromFuel`.

> relative_to is something that we should investigate, too. Though, I think this will be a separate PR and I would cross that bridge once we get there.

Definitely. From my side it does not have a great priority, even if it'd be nice to have.

Reply:
> move this code to a new `World::Impl::insertModel`

Actually, this doesn't work trivially, because `World::insertModel` calls `World::modelNames` which depends on attributes only available to `World` but not available to `World::Impl`, namely `ignition::gazebo::EntityComponentManager* m_ecm`.

Are you happy with me creating a private function inside of `World`, e.g. `World::insertModel(sdf::Root ...)` which acts as the shared base for the methods discussed above? If not, how would you like to have the code structured?

Reply:
A private method is a possible solution, even though I'm not yet sure I understood its necessity. I quickly tried locally, and the following signature inside `World::Impl` seems working fine:

```cpp
    bool insertModel(const std::shared_ptr<sdf::Root>& root,
                     const core::Pose& pose,
                     const std::string& overrideModelName,
                     World& world) // Note: this is a scenario::gazebo::World, 
                                   //       that has m_ecm attribute
```

Am I missing any detail?

Reply:
Yep, doing this can also work. It just seems odd to pass the outer class into a nested inner class because the inner needs access to outer's attributes. Having a private method in outer is likely better/cleaner software engineering. 

It depends a bit on what the `*::Impl` inner classes are supposed to be. I haven't quite worked this out yet, which is why I asked how the code should be structured. Are they meant to be structs that capture data that belongs together, or are they a means of hiding private attributes and methods?

Reply:
I agree with you, what you described is the only drawback of the [pimpl idiom](https://cpppatterns.com/patterns/pimpl.html) that we use throughout the project. I personally use this idiom pretty much everywhere and organization-wise too, this is not the first time this problem occurs in our projects. However, I think that the benefits overcome the drawbacks. 

The other workaround, in my opinion even worse but it depends how many time is necessary, is passing a pointer to the public class to the `Impl` constructor.

Reply:
Thanks for the added detail 💯 I'll go ahead with the signature you suggested then.